include/benchmark/benchmark_api.h

```diff
@@ -609,11 +756,15 @@ class Fixture: public internal::Benchmark {
 
     virtual void Run(State& st) {
       this->SetUp(st);
+      this->InitState(st);
       this->BenchmarkCase(st);
+      this->TerminateState(st);
       this->TearDown(st);
     }
 
     virtual void SetUp(const State&) {}
+    virtual void InitState(State &) {}
```

Comment:
As a usage example, here's a PAPI fixture class. Note that the counters are started and stopped in these new Fixture methods, and not inside the actual benchmark.

```
#include "benchmark/benchmark_api.h"                                                                                     
                                                                                                                         
#include "papi.h"                                                                                                        
                                                                                                                         
#include <random>                                                                                                        
#include <algorithm>                                                                                                     
#include <iostream>                                                                                                      
                                                                                                                         
                                                                                                                         
class PAPIFixture : public ::benchmark::Fixture {                                                                        
public:                                                                                                                  
                                                                                                                         
  void SetUp(const benchmark::State& st) {                                                                               
    values.resize(st.range_x());                                                                                         
    sequence.resize(st.range_x());                                                                                       
    int count = 0;                                                                                                       
    for(int &i : sequence) i = count++;                                                                                  
    std::random_device rng;                                                                                              
    std::mt19937 urng(rng());                                                                                            
    std::shuffle(sequence.begin(), sequence.end(), urng);
  }

  void InitState(benchmark::State &) {
    if(PAPI_start_counters(&event_ids[0], (int)event_ids.size()) != PAPI_OK) {
        std::cerr << "could not start PAPI counters\n";
    }
  }
  void TerminateState(benchmark::State &st) {
    if(PAPI_stop_counters(&events[0], (int)events.size()) != PAPI_OK) {
        std::cerr << "could not get PAPI counters\n";
    }
    for(int i = 0, s = (int)events.size(); i < s; i++) {
      st.SetCounter(event_names[i], (double)events[i]);
    }
  }

  std::vector< int          > event_ids   = { PAPI_L1_DCM ,  PAPI_L2_DCM ,/*  PAPI_L3_DCM ,  PAPI_MEM_SCY ,*/  PAPI_FP_INS };
  std::vector< const char * > event_names = {"PAPI_L1_DCM", "PAPI_L2_DCM",/* "PAPI_L3_DCM", "PAPI_MEM_SCY",*/ "PAPI_FP_INS"};
  std::vector< long_long    > events      = { 0           ,  0           ,/*  0           ,  0            ,*/  0           };


  //--------------------------------------------------------------

  float reduce_linear() const {
    float sum = 0.;
    for(size_t i = 0, s = values.size(); i < s; ++i) {
      sum += values[i];
    }
    return sum;
  }
  float reduce_jumps() const {
    float sum = 0.;
    for(size_t i = 0, s = values.size(); i < s; ++i) {
        sum += values[sequence[i]];
    }
    return sum;
  }

  std::vector< int > values;
  std::vector< int > sequence;

};



BENCHMARK_DEFINE_F(PAPIFixture, ReduceLinear)(benchmark::State& st) {
  float sum = 0.;
  while (st.KeepRunning()) {
    benchmark::DoNotOptimize(sum);
    sum = reduce_linear();
  }
}
BENCHMARK_DEFINE_F(PAPIFixture, ReduceJumps)(benchmark::State& st) {
  float sum = 0.;
  while (st.KeepRunning()) {
    benchmark::DoNotOptimize(sum);
    sum = reduce_jumps();
  }
}
BENCHMARK_REGISTER_F(PAPIFixture, ReduceLinear)->Arg(1<<6 );
BENCHMARK_REGISTER_F(PAPIFixture, ReduceJumps )->Arg(1<<6 );
BENCHMARK_REGISTER_F(PAPIFixture, ReduceLinear)->Arg(1<<8 );
BENCHMARK_REGISTER_F(PAPIFixture, ReduceJumps )->Arg(1<<8 );
BENCHMARK_REGISTER_F(PAPIFixture, ReduceLinear)->Arg(1<<10);
BENCHMARK_REGISTER_F(PAPIFixture, ReduceJumps )->Arg(1<<10);
BENCHMARK_REGISTER_F(PAPIFixture, ReduceLinear)->Arg(1<<12);
BENCHMARK_REGISTER_F(PAPIFixture, ReduceJumps )->Arg(1<<12);
BENCHMARK_REGISTER_F(PAPIFixture, ReduceLinear)->Arg(1<<20);
BENCHMARK_REGISTER_F(PAPIFixture, ReduceJumps )->Arg(1<<20);
BENCHMARK_REGISTER_F(PAPIFixture, ReduceLinear)->Arg(1<<24);
BENCHMARK_REGISTER_F(PAPIFixture, ReduceJumps )->Arg(1<<24);

BENCHMARK_MAIN()
```